File: android/app/src/main/java/bjc/shoperp/service/restful/ServiceContainer.java
package bjc.shoperp.service.restful;

import java.lang.reflect.Type;
import java.util.ArrayList;

import static java.lang.System.in;

/**
 * Created by hcq on 2018/2/15.
 */
public class ServiceContainer {

    public static String ServerAddress = null;

    public static String AccessToken ="";

    static ArrayList<Object> services = new ArrayList<Object>();

    public static <T > T  GetService(Class<T> c) throws Exception {
        String cname=c.getName();
        if(cname.endsWith("Service")==false ){
            throw new Exception("类型："+cname+"名称结尾不为Service无法创建");
        }
        for (Object obj : services) {
            if(obj.getClass()==c){
                return (T)obj;
            }
        }
        T t =c.newInstance();
        services.add(t);
        return  t;
    }
}

File: net/ShopErp.App/Service/Restful/ServiceContainer.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;


namespace ShopErp.App.Service.Restful
{
    public class ServiceContainer
    {
        public static string ServerAddress = null;

        public static string AccessToken { get; set; }

        static List<object> services = new List<object>();

        static ServiceContainer()
        {
            Type[] types = typeof(ServiceContainer).Assembly.GetTypes();

            foreach (Type t in types)
            {
                if (t.BaseType != null && t.Namespace == "ShopErp.App.Service.Restful" && t.Name.EndsWith("Service"))
                {
                    services.Add(Activator.CreateInstance(t));
                }
            }
        }

        public static T GetService<T>()
        {
            object obj = services.FirstOrDefault(o => o.GetType() == typeof(T) || o.GetType().GetInterface(typeof(T).Name, true) != null);
            if (obj == null)
            {
                throw new Exception("未找到指定的服务类型:" + typeof(T).FullName);
            }
            return (T)obj;
        }
    }
}